osu.Server.BeatmapSubmission/BeatmapSubmissionController.cs

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using MySqlConnector;
+using osu.Framework.Extensions;
+using osu.Game.IO.Archives;
+using osu.Server.BeatmapSubmission.Authentication;
+using osu.Server.BeatmapSubmission.Models;
+using osu.Server.BeatmapSubmission.Models.API.Requests;
+using osu.Server.BeatmapSubmission.Models.API.Responses;
+using osu.Server.BeatmapSubmission.Models.Database;
+using osu.Server.BeatmapSubmission.Services;
+using osu.Server.QueueProcessor;
+
+namespace osu.Server.BeatmapSubmission
+{
+    [Authorize]
+    public class BeatmapSubmissionController : Controller
+    {
+        private readonly IBeatmapStorage beatmapStorage;
+        private readonly BeatmapPackagePatcher patcher;
+
+        public BeatmapSubmissionController(IBeatmapStorage beatmapStorage, BeatmapPackagePatcher patcher)
+        {
+            this.beatmapStorage = beatmapStorage;
+            this.patcher = patcher;
+        }
+
+        // TODO: accept pending/WIP user choice somewhere
+
+        /// <summary>
+        /// Create a new beatmap set, or add/remove beatmaps from an existing beatmap set
+        /// </summary>
+        /// <response code="200">The requested changes have been applied.</response>
+        /// <response code="403">The user is not allowed to modify this beatmap set.</response>
+        /// <response code="404">The request specified a beatmap set that does not yet exist.</response>
+        /// <response code="429">The request was incorrectly formed. Check returned error for details.</response>
```

Comment:
a bit late but 429 is "too many requests", I think you mean 422 😄

Reply:
Yeah I keep messing the two up for whatever reason. Will fix in the next one in chain, it'll be dealing with these invariant exceptions et cetera anyways.